website/files/models/base.py

```diff
@@ -0,0 +1,588 @@
+    path = fields.StringField(required=True)
+    materialized_path = fields.StringField(required=True)
+
+    @property
+    def deep_url(self):
+        return self.wrapped().deep_url
+
+    def wrapped(self):
+        """Wrap self in a FileNode subclass
+        """
+        return PROVIDER_MAP[self.provider][int(self.is_file)](self)
+
+    def get_guid(self, create=False):
+        """Attempt to find a Guid that points to this object.
+        One will be created if requested.
+        :rtype: Guid
+        """
+        try:
+            return Guid.find_one(Q('referent', 'eq', self))
+        except NoResultsFound:
+            if not create:
+                return None
+        return Guid.generate(self)
+
+
+class FileNodeMeta(type):
+    """Keeps track of subclasses of the ``FileNode`` object
+    Inserts all into the PROVIDER_MAP following the pattern:
+    {
+        provider: [ProviderFolder, ProviderFile, ProviderFileNode]
+    }
+    """
+
+    def __init__(cls, name, bases, dct):
+        super(FileNodeMeta, cls).__init__(name, bases, dct)
+        if hasattr(cls, 'provider'):
+            cls_map = PROVIDER_MAP.setdefault(cls.provider, [None, None, None])
+            index = int(getattr(cls, 'is_file', 2))
+
+            if cls_map[index] is not None:
+                raise ValueError('Conflicting providers')
+
+            cls_map[index] = cls
+
+
+class FileNode(object):
+    """The base class for the entire files storage system.
+    Use for querying on all files and folders in the database
+    """
+    FOLDER, FILE, ANY = 0, 1, 2
+
+    __metaclass__ = FileNodeMeta
+
+    @classmethod
+    def create(cls, **kwargs):
+        """A layer of abstraction around the creation of FileNodes.
+        Provides hook in points for subclasses
+        This is used only for GUID creation.
+        """
+        assert hasattr(cls, 'is_file') and hasattr(cls, 'provider'), 'Must have is_file and provider to call create'
+        kwargs['is_file'] = cls.is_file
+        kwargs['provider'] = cls.provider
+        return cls(**kwargs)
+
+    @classmethod
+    def get_or_create(cls, node, path):
+        """Tries to find a FileNode with node and path
+        See FileNode.create
+        """
+        path = '/' + path.lstrip('/')
+        try:
+            # Note: Possible race condition here
+            # Currently create then find is not super feasable as create would require a
+            # call to save which we choose not to call to avoid filling  the database
+            # with notfound/googlebot files/url. Raising 404 errors may roll back the transaction however
+            return cls.find_one(
+                Q('node', 'eq', node) &
+                Q('path', 'eq', path)
+            )
+        except NoResultsFound:
+            return cls.create(node=node, path=path)
+
+    @classmethod
+    def resolve_class(cls, provider, _type=2):
+        """Resolves a provider and type to the appropriate subclass.
```

Comment:
Hate to nitpick, but PEP 257 recommends using the imperative ("Resolve a provider").

I dig the usage examples, though.